src/main/java/patternverifiers/SingletonVerifier.java

```diff
@@ -1,17 +1,312 @@
+import com.github.javaparser.ast.body.ConstructorDeclaration;
+import com.github.javaparser.ast.body.FieldDeclaration;
+import com.github.javaparser.ast.body.MethodDeclaration;
+import com.github.javaparser.ast.expr.BinaryExpr;
+import com.github.javaparser.ast.expr.MethodCallExpr;
+import com.github.javaparser.ast.expr.ObjectCreationExpr;
+import com.github.javaparser.ast.stmt.IfStmt;
+
+import base.VariableReader;
 
 /**
  * A verifier for the singleton pattern.
  */
 public class SingletonVerifier implements IPatternVerifier {
 
+    private final transient List<ConstructorDeclaration> constDeclList = new ArrayList<>();
+    private transient FieldDeclaration instanceVar;
+
     public SingletonVerifier() {
     }
 
     @Override
     public Feedback verify(CompilationUnit compUnit) {
-        throw new UnsupportedOperationException();
+        boolean isValid =
+            callsConstructor(compUnit) && hasAStaticInstance(compUnit) && onlyInstantiatedIfNull(
+            compUnit) && hasPrivateConstructor(compUnit);
+
+        return new Feedback(isValid);
+    }
+
+    /**
+     * Method for checking if all the constructors in a Java class are private.
+     *
+     * @param compUnit The CompilationUnit representing the Java class to look at
+     *
+     * @return True iff all constructors are private
+     */
+    public boolean hasPrivateConstructor(CompilationUnit compUnit) {
+        List<Boolean> listOfIsPrivate = new ArrayList<>();
+        boolean isPrivConstr = true;
+
+        // Finds and checks if the constructors are private or not.
+        compUnit.findAll(ConstructorDeclaration.class).forEach(constructor -> {
+            if (constructor.isPrivate()) {
+                listOfIsPrivate.add(true);
+            } else {
+                listOfIsPrivate.add(false);
+            }
+        });
+
+        // Takes all boolean values and ANDs them to one value.
+        for (Boolean bool : listOfIsPrivate) {
+            isPrivConstr &= bool;
+        }
+
+        return isPrivConstr;
+    }
+
+    /**
+     * Method for declaring if a java class holds a field variable of a static instance
+     * https://stackoverflow .com/questions/53300710/how-to-parse-inner-class-from-java-source-code
+     * might help solving a check for inner classes
+     *
+     * @param compUnit The CompilationUnit representing the java class to look at
+     *
+     * @return True iff the java class holds a field variable with a static modifier of the same
+     *     type as the class itself (eg. static SingletonVerifier sv;)
+     */
+    public boolean hasAStaticInstance(CompilationUnit compUnit) {
+        boolean isStatic = false;
+        boolean isPrivate = false;
+        boolean isInstantiated = false;
+
+        for (FieldDeclaration fieldDeclaration : VariableReader.readVariables(
+            compUnit)) {      // For each FieldDeclaration in the java file
+            if (fieldDeclaration.getVariables().get(0).getType().toString().equals(
```

Comment:
When the chain is this long is is much better to abstract it into a separate function where the purpose can be explained in the method name

Reply:
Could be, but I doubt this would be reused, and the chain is long due to strange return types etc from javaparser. Haven't resolved this for now but could definately do if wanted

Reply:
It is not for the case of reuse that I recommend creating a function, but rather adding clarity and less clutter to the actual logic of this class.